test/testground/network/sync.go

```diff
@@ -0,0 +1,218 @@
+	GroupID        string            `json:"group_id"`
+	GlobalSequence int64             `json:"global_sequence"`
+	CmtConfig      *tmconfig.Config  `json:"cmt_config"`
+	AppConfig      *srvconfig.Config `json:"app_config"`
+}
+
+func peerIDs(nodes []RoleConfig) []string {
+	peerIDs := make([]string, 0, len(nodes))
+	for _, nodeCfg := range nodes {
+		peerIDs = append(peerIDs, nodeCfg.PeerID)
+	}
+	return peerIDs
+}
+
+func searchNodes(nodes []RoleConfig, globalSeq int64) (RoleConfig, bool) {
+	for _, node := range nodes {
+		if node.GlobalSequence == globalSeq {
+			return node, true
+		}
+	}
+	return RoleConfig{}, false
+}
+
+func PublishNodeConfigs(ctx context.Context, initCtx *run.InitContext, nodes []RoleConfig) error {
+	for _, node := range nodes {
+		_, err := initCtx.SyncClient.Publish(ctx, MetaConfigTopic, node)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func DownloadNodeConfigs(ctx context.Context, runenv *runtime.RunEnv, initCtx *run.InitContext) ([]RoleConfig, error) {
+	return DownloadSync(ctx, initCtx, MetaConfigTopic, RoleConfig{}, runenv.TestInstanceCount)
+}
+
+// GenesisWrapper is a simple struct wrapper that makes it easier testground to
+// properly serialize and distribute the genesis file.
+type GenesisWrapper struct {
+	Part    int    `json:"part"`
+	Genesis string `json:"genesis"`
+}
+
+// PublishGenesis publishes the genesis to the sync service. It splits the
+// genesis into 10 parts and publishes each part separately. This gets around the
+// 32Kb limit the underlying websocket has
+func PublishGenesis(ctx context.Context, initCtx *run.InitContext, gen *coretypes.GenesisDoc) error {
+	genBytes, err := cmtjson.Marshal(gen)
+	if err != nil {
+		return err
+	}
+
+	wrappers := make([]GenesisWrapper, 0, wrapperParts)
+	partSize := len(genBytes) / wrapperParts
+
+	for i := 0; i < wrapperParts; i++ {
+		start := i * partSize
+		end := start + partSize
+		if i == wrapperParts-1 {
+			end = len(genBytes)
+		}
+
+		wrappers = append(wrappers, GenesisWrapper{
+			Part:    i,
+			Genesis: hex.EncodeToString(genBytes[start:end]),
+		})
+	}
+
+	for _, wrapper := range wrappers {
+		_, err = initCtx.SyncClient.Publish(ctx, GenesisTopic, wrapper)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func DownloadGenesis(ctx context.Context, initCtx *run.InitContext) (json.RawMessage, error) {
+	rawGens, err := DownloadSync(ctx, initCtx, GenesisTopic, GenesisWrapper{}, wrapperParts)
+	if err != nil {
+		return nil, err
+	}
+	// sort the genesis parts by their part number
+	sort.Slice(rawGens, func(i, j int) bool {
+		return rawGens[i].Part < rawGens[j].Part
+	})
+	var genesis []byte
+	for _, rawGen := range rawGens {
+		bz, err := hex.DecodeString(rawGen.Genesis)
+		if err != nil {
+			return nil, err
+		}
+		genesis = append(genesis, bz...)
+	}
+	return genesis, nil
+}
+
+func DownloadSync[T any](ctx context.Context, initCtx *run.InitContext, topic *sync.Topic, t T, count int) ([]T, error) {
```

Comment:
A func doc would be very helpful.

Reply:
261214a5adbc10044fac9c6c3b4b07ba18c2aea0